### ISSUE TITLE
Terragrunt cannot assume role (using Atlantis)

### ISSUE THREAD

**Issue by michal-adamkiewicz** [gruntwork-io/terragrunt#1918]
### TL;DR;
IAM role assume works perfectly fine from a shell inside docker. It works also using an aws cli command inside Atlantis. But it throws an error when I want to use Terragrunt "way" of assuming - **--terragrunt-iam-role** & **TERRAGRUNT_IAM_ROLE** env.

### Details:

In my current project, I had to incorporate Terragrunt into Atlantis (I've done this before many times using K8s and working on multi-account architecture) using ECS on AWS. To be short in this introduction - I already have ECS (Fargate) hosting my docker, but unfortunately seems that I'm struggling with issues that Terragrunt cannot assume my IAM roles.

I start my Fargate task with the task execution role and task role attached. Then by the built-in feature in Terragrunt for role-assume, I would like to switch my role and context into another AWS account.

IAM role for ECS task (let's called: **task_role** or source role) has attached permissions to assume next role (let's called it **automation_role_account_B** or just target role):

**arn:aws:iam::12345:role/task_role**

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "sts:AssumeRole"
            ],
            "Resource": [
                "arn:aws:iam::123456:role/automation_role_account_B"
            ]
        }
    ]
}
```

Where, a "target role" of course has adjusted trust relationship policy to allow being assumed be previous one:

**arn:aws:iam::123456:role/automation_role_account_B**

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::12345:role/task-role"
      },
      "Action": "sts:AssumeRole"
    }
  ]
```

From AWS perspective everything looks fine....
Now fun starts as I build my docker with Atlantis + AWS CLI + Terragrunt etc.
Inside Atlantis I have the opportunity to execute custom commands/workflows where I use Terragrunt:

```
workflows:
  terragrunt-test:
    plan:
      steps:
      - env:
          name: TERRAGRUNT_IAM_ROLE
          value: arn:aws:iam::123456:role/automation_role_account_B
      - env:
          name: TF_CLI_ARGS
          value: "-no-color"
      - env:
          name: TERRAGRUNT_TFPATH
          command: 'echo "/home/atlantis/.atlantis/bin/terraform$ATLANTIS_TERRAFORM_VERSION"'
      - run: aws sts get-caller-identity
      - run: terragrunt plan -no-color

```

I'm using in my use case variable **TERRAGRUNT_IAM_ROLE** to provide an IAM role I would like to assume, but the same (unsuccessful) results I have got when I used a dedicated terragrung flag: **terragrunt plan --terragrunt-iam-role ...**

The error that I have got clear claims about 403 error (access denied):

```
[terragrunt] 2021/11/17 23:45:02 Terraform version: 0.14.9
[terragrunt] 2021/11/17 23:45:02 Reading Terragrunt config file at /home/atlantis/.atlantis/repos/test/we-devops/atlantis-example/3/default/terragrunt/my-test/eu-west-1/apps/test-app/terragrunt.hcl
[terragrunt] 2021/11/17 23:45:02 Running command: bash /home/atlantis/.atlantis/repos/test/we-devops/atlantis-example/3/default/terragrunt/../util/get_user_env.sh my-test/eu-west-1/apps/test-app
my-test[terragrunt] 2021/11/17 23:45:02 run_cmd output: [account-test]
[terragrunt] 2021/11/17 23:45:02 Assuming IAM role arn:aws:iam::123456:role/automation_role_account_B
[terragrunt] 2021/11/17 23:45:03 AccessDenied: User: arn:aws:sts::12345:assumed-role/task-role/decf22cd78b34e1ab9472d19e06d1a2f is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::123456:role/automation_role_account_B
	status code: 403, request id: 613250cc-58e1-4fd8-a76f-64230b004050
[terragrunt] 2021/11/17 23:45:03 Unable to determine underlying exit code, so Terragrunt will exit with error code 1

{
    "UserId": "AROARLQNH54TED56DS6AX:decf22cd78b34e1ab9472d19e06d1a2f",
    "Account": "12345",
    "Arn": "arn:aws:sts::12345:assumed-role/task-role/decf22cd78b34e1ab9472d19e06d1a2f"
}
```

But.... when I assume a role any other way, it works!.
I can for instance add one more line to Atlantis like this:

` - run: aws sts assume-role --role-arn arn:aws:iam::123456:role/automation_role_account_B --role-session-name cs
`
and this will work.
I connected to the Fargate container using EXEC functionality (https://aws.amazon.com/blogs/containers/new-using-amazon-ecs-exec-access-your-containers-fargate-ec2/) and get access to an interactive shell inside my container, where again I was able just successfully assume my role:

*aws sts get-caller-identity*
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/54705469/142302008-7b321caf-f7a4-4ceb-8ed2-67369be6856a.png">

*aws sts assume-role --role-arn arn:aws:iam::123456:role/automation_role_account_B --role-session-name cs*
<img width="1368" alt="image" src="https://user-images.githubusercontent.com/54705469/142302130-d143a8b8-bd6c-415e-b0b2-4fe3fef1aba9.png">

**Completly don't get it :/**

Versions of applications inside my docker:

PYTHON_VERSION=3.8.7-r1
AWS_CLI_VERSION=2.0.30
TERRAGRUNT_VERSION=v0.27.0 (but I have testes as well on the latest 0.35.10, and exactly the same error)
TERRAFORM: 0.14.9

**Comment by wrnu** [comment]
I observe the exact same behaviour when running terragrunt in a gitlab runner on an ec2 instance.

The ec2 instance role has permission to assume a role in another account but terragrunt fails with the same type of 403 while attempting to assume the cross account role in the AWS terraform provider.

In the same gitlab job running on the instance the AWS cli can assume the role using `aws sts assume-role`.

Because I am not using `iam_role` or `--terrangrunt-iam-role` but instead assuming the role with the terraform provider it seems like the issue may also exist in terraform. Or possibly the AWS Go SDK since both terraform and terragrunt use that.

**Comment by yorinasub17** [comment]
Does the role have restrictions on any of the following?

- Session duration. Terragrunt defaults to 3600 if the session duration isn't provided (https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-iam-assume-role-duration)
- Session name. Terragrunt will not configure a role session name if it isn't provided (https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-iam-assume-role-session-name)

**Comment by tehlers320** [comment]
I'm seeing this too, i believe its this block here https://github.com/gruntwork-io/terragrunt/blob/300e1e2a9dca27f524ff4bca9e6b665d8e4fa538/aws_helper/config.go#L128 the env vars attempt is made here: https://github.com/gruntwork-io/terragrunt/blob/300e1e2a9dca27f524ff4bca9e6b665d8e4fa538/aws_helper/config.go#L138

We are trying to chain assume roles but it breaks because the environment vars from the first assume role need to happen then a 2nd assume role.

**Comment by jvanceACX** [comment]
I encountered this issue when working with older infrastructure (deployed ~12 months ago on tf 0.13 + terragrunt 0.25) and while I can't guarantee this is all strictly necessary, what finally worked for me was:

1. reverting both terraform and terragrunt to the particular versions initially deployed with
2. setting `version  = "3.47.0"` in my aws provider block (deprecated but it works in .13)
3. terragrunt init
4. `rm .terraform/plugins/registry.terraform.io/-/aws`
5. `cp .terraform/plugins/registry.terraform.io/hashicorp/aws .terraform/plugins/registry.terraform.io/-/aws`

Those last couple of steps were because I could not figure out how to force init to download an older plugin for `-/aws` which is what tf .13 actually kept trying to use. I don't believe it's relevant for later versions of terraform.

Anyway. After all that, I was able to plan, apply, etc without the constant 403s. Unfortunately my scenario is specific enough and my solution broad/inelegant enough that it may be of no value in either identifying the source of this problem or helping anyone else work around their own issues.

**Comment by tehlers320** [comment]
@jvanceACX im not sure its the same issue, the block i highlighted is from terragrunt which seems to at least be where my issue is from. It doesnt even make it down to terraform before assuming the wrong role.

**Comment by juozasget** [comment]
Hello. I just wanted to leave a note that we stumbled upon what seems to be the same issue as described here.
We can't run terragrunt from GH actions when using a special github IAM role that is assumed via OIDC.

IAM role chaining happens and terraform/terragrunt can't assume the correct role. When I removed the AWS credentials and token environment variables and added static credential files - everything started to work.

To me it looks like it might be a terraform issue, rather than a terragrunt issue. Because the errors where present even when testing with terraform (in the debug environment).

**Comment by jhrr** [comment]
@juozasget did you ever manage to resolve this for OIDC runs? I'm having the same issue I think...

**Comment by juozasget** [comment]
@jhrr Hey! Late reply - but no, we just continued to use static credentials to workaround the issue. By now we are using IAM roles via OIDC + Identity Center for human users with dynamic credentials. So it would be nice to get this working with dynamic credentials also :)
